### PR TITLE
chore(site): align ESLint config to typescript-eslint's recommended-requiring-type-checking

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
From https://twitter.com/jsjoeio/status/1616168014172557314:

1. Has the config extend https://typescript-eslint.io/linting/configs#recommended-requiring-type-checking, not just https://typescript-eslint.io/linting/configs#recommended
2. Removes rule overrides that aren't necessary. Those are mostly:
    * Stylistic rules ([we strongly recommend using a separate dedicated formatter](https://typescript-eslint.io/linting/troubleshooting/formatting))
    * Rules now enabled by the config (e.g. [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises))
    * Rules that might have previously needed to be disabled, but are no longer enabled in the recommended configs (e.g. [`@typescript-eslint/camelcase`](https://typescript-eslint.io/rules/camelcase))
3. Fixed all immediately fixable rule complaints (commented inline)
4. Disabled the rest of the newly complaining rules in the ESLint config, with a TODO comment each

I'd definitely recommend filing those TODOs in your backlog and linking to them in the config file. Most of those rules can catch an instance of `any` or some other unfortunate practice sneaking into your app. Though, note that the ones around template string concatenation might not have as many (or any) reports after https://github.com/typescript-eslint/typescript-eslint/discussions/6014 lands.

### CODE REVIEW

**Review by kylecarbs — APPROVED** [review]
Thanks for the contribution! ✨

**Comment by JoshuaKGoldberg** [comment]
Wait wait, it's still a draft and I still have to fix or disable the lint complaints. There's much more to come 😄

**Comment by JoshuaKGoldberg** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by kylecarbs** [comment]
@JoshuaKGoldberg I figured, but still agree with the thread linked. I'll re-review once all the changes are made!

File: site/src/api/api.ts [hunk 1]
```diff
@@ -20,8 +20,7 @@ export const hardCodedCSRFCookie = (): string => {
 export const withDefaultFeatures = (
   fs: Partial<TypesGen.Entitlements["features"]>,
 ): TypesGen.Entitlements["features"] => {
-  for (const k in TypesGen.FeatureNames) {
-    const feature = k as TypesGen.FeatureName
+  for (const feature of TypesGen.FeatureNames) {
```
Comment (JoshuaKGoldberg) [inline site/src/api/api.ts:23]: https://typescript-eslint.io/rules/no-for-in-array

> A for-in loop (`for (var i in o)`) iterates over the properties of an Object. While it is legal to use for-in loops with array types, it is not common. for-in will iterate over the indices of the array as strings, omitting any "holes" in the array.

File: site/src/api/api.ts [hunk 2]
```diff
@@ -152,8 +151,7 @@ export const getTokens = async (): Promise<TypesGen.APIKey[]> => {
 }
 
 export const deleteAPIKey = async (keyId: string): Promise<void> => {
-  const response = await axios.delete("/api/v2/users/me/keys/" + keyId)
-  return response.data
+  await axios.delete("/api/v2/users/me/keys/" + keyId)
```
Comment (JoshuaKGoldberg) [inline site/src/api/api.ts:154]: This function indicates it returns `Promise<void>`, so any returned data is ignored. It's violating [`@typescript-eslint/no-unsafe-return`](https://typescript-eslint.io/rules/no-unsafe-return) anyway so I figured I'd clean it up. 🙂

File: site/src/components/LicenseBanner/LicenseBanner.test.tsx [hunk 3]
```diff
@@ -9,8 +9,8 @@ import { Language } from "./LicenseBannerView"
 describe("LicenseBanner", () => {
   it("does not show when there are no warnings", async () => {
     render(<LicenseBanner />)
-    const bannerPillSingular = await screen.queryByText(Language.licenseIssue)
-    const bannerPillPlural = await screen.queryByText(Language.licenseIssues(2))
+    const bannerPillSingular = screen.queryByText(Language.licenseIssue)
+    const bannerPillPlural = screen.queryByText(Language.licenseIssues(2))
```
Comment (JoshuaKGoldberg) [inline site/src/components/LicenseBanner/LicenseBanner.test.tsx:13]: `screen.queryByText`, `screen.getByRole`, etc. don't return Promises. There's no need to `await` them.

[`@typescript-eslint/await-thenable`](https://typescript-eslint.io/rules/await-thenable)

File: site/src/xServices/users/searchUserXService.ts [hunk 4]
```diff
@@ -51,9 +51,7 @@ export const searchUserMachine = createMachine(
   {
     services: {
       searchUsers: async (_, { query }) =>
-        await (
-          await getUsers(queryToFilter(query))
-        ).users,
+        (await getUsers(queryToFilter(query))).users,
```
Comment (JoshuaKGoldberg) [inline site/src/xServices/users/searchUserXService.ts:54]: This also is an `await` of something (`.users`) that is not async.

**Comment by BrunoQuaresma** [comment]
@JoshuaKGoldberg thx for your contribution! I'm going to solve the conflict and merge it.

**Comment by JoshuaKGoldberg** [comment]
...huh, I swear I used to know where that option was. Fixed!

**Pull request merged**